### PR TITLE
replace Unknowns with free type variables in string representations

### PR DESCRIPTION
**Pull request by davidchambers**
Commit message:

> Currently `???` is used for both Inconsistent and Unknown:
>
>     ['foo', true, 42] :: Array ???
>     []                :: Array ???
>
> This commit visually differentiates the two types:
>
>     ['foo', true, 42] :: Array ???
>     []                :: Array a
>
> The type variable is chosen to avoid collisions with variables in the corresponding type signature.

This pull request was motivated by comments in #160.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -2095,11 +2096,38 @@
...
+    for (var code = a; code <= z; code += 1) {
+      chars.push(String.fromCharCode(code));
+    }
+    return chars;
+  }());
+
+  //  showType :: TypeInfo -> Type -> String
+  function showType(typeInfo) {
+    return function(t) {
+      var names = Z.chain(typeVarNames, typeInfo.types);
```
Comment (Avaq) [inline index.js]: You can hoist this computation so that it will not be executed for every type.

File: index.js [hunk 2]
```diff
@@ -2095,11 +2096,38 @@
...
+      var $free = typeVarChars.filter(function(name) {
+        return names.indexOf(name) < 0;
+      });
+      return unless(
+        t.type === FUNCTION || t.type === RECORD || isEmpty(t.keys),
+        stripOutermostParens,
+        String(t).replace(/\bUnknown\b/g, function() { return $free.shift(); })
```
Comment (Avaq) [inline index.js]: If you use `$free[idx++]` instead, you can also hoist the computation for `$free`.

File: index.js [hunk 3]
```diff
@@ -2109,17 +2137,24 @@
...
+    typeInfo,       // :: TypeInfo
+    values,         // :: Array Any
+    pos             // :: Integer
+  ) {
     return String(pos) + ')  ' + Z.map(function(x) {
       var types = determineActualTypesLoose(env, [x]);
-      return Z.toString(x) + ' :: ' + Z.map(showType, types).join(', ');
+      return Z.toString(x) +
+             ' :: ' +
+             Z.map(showType(typeInfo), types).join(', ');
```
Comment (Avaq) [inline index.js]: You could assign the result of `showType(typeInfo)` to a variable to make better use of the optimizations proposed before.

**Comment by davidchambers** [comment]
I've made some changes based on your feedback, Aldwin. Please have another look. :)

File: index.js [hunk 4]
```diff
@@ -2095,11 +2096,29 @@
...
+    var names = Z.chain(typeVarNames, typeInfo.types);
+    return function(t) {
+      var code = 'a'.charCodeAt(0);
+      return unless(
+        t.type === FUNCTION || t.type === RECORD || isEmpty(t.keys),
+        stripOutermostParens,
+        String(t).replace(/\bUnknown\b/g, function() {
+          do var name = String.fromCharCode(code++);
+          while (names.indexOf(name) >= 0);
+          return name;
```
Comment (davidchambers) [inline index.js]: The rarely seen `do...while`!

**Pull request merged**